kernel/cost.cc

```diff
@@ -0,0 +1,226 @@
+#include "kernel/cost.h"
+
+USING_YOSYS_NAMESPACE
+
+unsigned int CellCosts::get(RTLIL::Module *mod)
+{
+	if (mod_cost_cache_.count(mod->name))
+		return mod_cost_cache_.at(mod->name);
+
+	unsigned int module_cost = 1;
+	for (auto c : mod->cells()) {
+		unsigned int new_cost = module_cost + get(c);
+		module_cost = new_cost >= module_cost ? new_cost : INT_MAX;
+	}
+
+	mod_cost_cache_[mod->name] = module_cost;
+	return module_cost;
+}
+
+static unsigned int y_coef(RTLIL::IdString type)
+{
+	// clang-format off
+    if (// equality
+        type == ID($bweqx) ||
+        type == ID($nex) ||
+        type == ID($eqx) ||
+        // basic logic
+        type == ID($and) ||
+        type == ID($or) ||
+        type == ID($xor) ||
+        type == ID($xnor) ||
+        type == ID($not) ||
+        // mux
+        type == ID($bwmux) ||
+        type == ID($mux) ||
+        type == ID($demux) ||
+        // others
+        type == ID($tribuf)) {
+        return 1;
+    } else if (type == ID($neg)) {
+        return 4;
+    } else if (type == ID($fa)) {
+        return 5;
+    } else if (// multi-bit adders
+        type == ID($add) ||
+        type == ID($sub) ||
+        type == ID($alu)) {
+        return 8;
+    } else if (// left shift
+        type == ID($shl) ||
+        type == ID($sshl)) {
+        return 10;
+    }
+	// clang-format on
+	return 0;
+}
+
+static unsigned int max_inp_coef(RTLIL::IdString type)
+{
+	// clang-format off
+    if (// binop reduce
+        type == ID($reduce_and) ||
+        type == ID($reduce_or) ||
+        type == ID($reduce_xor) ||
+        type == ID($reduce_xnor) ||
+        type == ID($reduce_bool) ||
+        // others
+        type == ID($logic_not) ||
+        type == ID($pmux) ||
+        type == ID($bmux)) {
+        return 1;
+    } else if (// equality
+        type == ID($eq) ||
+        type == ID($ne) ||
+        // logic
+        type == ID($logic_and) ||
+        type == ID($logic_or)) {
+        return 2;
+    } else if (type == ID($lcu)) {
+        return 5;
+    } else if (// comparison
+        type == ID($lt) ||
+        type == ID($le) ||
+        type == ID($ge) ||
+        type == ID($gt) ||
+        // others
+        type == ID($sop)) {
+        return 6;
+	}
+	// clang-format on
+	return 0;
+}
+
+static unsigned int sum_coef(RTLIL::IdString type)
+{
+	// clang-format off
+    if (// right shift
+        type == ID($shr) ||
+        type == ID($sshr)) {
+        return 4;
+    } else if (// shift
+        type == ID($shift) ||
+        type == ID($shiftx)) {
+        return 8;
```

Comment:
Why is this twice the cost of `$shr`? I'm very unconvinced that the cost model is sound.

Reply:
I'll inspect some techmapped examples of these gates on Monday

Reply:
run `read_rtlil shift.il; techmap; opt; stat` on this shift.il:
```
module \shr
  wire width 10 input 1 \A
  wire width 10 input 2 \B
  wire width 10 output 3 \Y
  cell $shr \UUT
    parameter \A_SIGNED 1
    parameter \A_WIDTH 10
    parameter \B_SIGNED 1
    parameter \B_WIDTH 10
    parameter \Y_WIDTH 10
    connect \A \A
    connect \B \B
    connect \Y \Y
  end
end

module \shift
  wire width 10 input 1 \A
  wire width 10 input 2 \B
  wire width 10 output 3 \Y
  cell $shift \UUT
    parameter \A_SIGNED 1
    parameter \A_WIDTH 10
    parameter \B_SIGNED 1
    parameter \B_WIDTH 10
    parameter \Y_WIDTH 10
    connect \A \A
    connect \B \B
    connect \Y \Y
  end
end
```
yields
```
=== shift ===
   Number of cells:                122
=== shr ===
   Number of cells:                 64
```
Which shows that 2x factor. To the extent in which this yosys behavior is correct, so is the model